core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+                    }
+                    block.commitment = commitment;
+                    if block.l1_batch_number % 100 == 0 {
+                        tracing::info!("Finished processing l1_batch number {}, last processed priority tx: {}",
+                            block.l1_batch_number, last_processed_priority_tx - 1);
+                    }
+                    result.push(block)
+                }
+            }
+
+            if filter_to_block == end_block {
+                if let Some(last_block) = result.last() {
+                    tracing::info!(
+                        "Finished processing L1 data, last processed l1 batch: {}, \
+                    last processed priority tx id: {}",
+                        last_block.l1_batch_number,
+                        last_processed_priority_tx - 1
+                    );
+                }
+                break;
+            }
+            current_block = filter_to_block + 1;
+        }
+        return result;
+    }
+
+    async fn process_tx_data(
+        commit_functions: &[Function],
+        blob_client: &Arc<dyn BlobClient>,
+        tx: Transaction,
+        protocol_versioning: &ProtocolVersioning,
+    ) -> Result<Vec<CommitBlock>, ParseError> {
+        let block_number = tx.block_number.unwrap().as_u64();
+        loop {
+            match parse_calldata(
+                protocol_versioning,
+                block_number,
+                &commit_functions,
+                &tx.input.0,
+                &blob_client,
+            )
+            .await
+            {
+                Ok(blks) => break Ok(blks),
+                Err(e) => {
+                    match e.clone() {
+                        ParseError::BlobStorageError(err) => {
+                            tracing::error!("Blob storage error {err}");
+                        }
+                        ParseError::BlobFormatError(data, inner) => {
+                            tracing::error!("Cannot parse {}: {}", data, inner);
+                        }
+                        _ => {
+                            tracing::error!("Failed to parse calldata: {e}, encountered on block: {block_number}");
+                        }
+                    }
+                    return Err(e);
+                }
+            }
+        }
+    }
+
+    /// Get a specified transaction on L1 by its hash.
+    async fn get_transaction_by_hash(
+        eth_client: &Box<DynClient<L1>>,
+        hash: H256,
+    ) -> Result<Transaction> {
+        match L1Fetcher::retry_call(
+            || L1Fetcher::query_client(eth_client).get_tx(hash),
+            L1FetchError::GetTx,
+        )
+        .await
+        {
+            Ok(Some(tx)) => Ok(tx),
+            Ok(None) => Err(anyhow!("unable to find transaction with hash: {}", hash)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn query_client(eth_client: &Box<DynClient<L1>>) -> &dyn EthInterface {
+        eth_client
```

Comment:
this function does explicit casting to an interface. Does rust compiler have problems without it?